text/0000-hybrid-tx-building.md

```diff
@@ -0,0 +1,281 @@
+
+- Title: `hybrid-tx-building`
+- Authors: [@lehnberg](mailto:daniel.lehnberg@protonmail.com)
+- Start date: `Mar 19, 2020`
+- RFC PR: Edit if merged: [mimblewimble/grin-rfcs#0000](https://github.com/mimblewimble/grin-rfcs/pull/0000) 
+- Tracking issue: [Edit if merged with link to tracking github issue]
+
+---
+
+# Summary
+[summary]: #summary
+
+This RFC establishes a baseline for transaction building in Grin between users and services, through the introduction of a new 'hybrid' method. Transactions are initiated used two standardized formats, **QR code** and **blob**, and can then be completed over Tor or clearnet using https.
```

Comment:
Keeping https will just mean users continue to use MITM services like ngrok, grinplusplus.com, and hedwig.im.

By offering QR code and blob, you've now switched from a truly standardized format (JSON - currently used for all of our slates), to 2 different "standardized" formats.

Reply:
Thanks for the review.

> Keeping https will just mean users continue to use MITM services

We cannot prevent users from running third party services, so that's definitely possible, but also unrelated to the hybrid transaction method that is proposed here.

What do you suggest? Not keeping https? In favour of Tor-only, I assume?

If so:
- How do services/users in regions where Tor is blocked transact?
- How do services/users that are prohibited culturally/legally from running Tor transact?
- How do mobile wallets transact?
- How does the Grin network transact if/when Tor is down or unavailable?

To me, this would be like shooting ourselves in the foot when it comes to increasing adoption and ease of use, and it introduces a critical external dependency on a third party. The added friction might also lead to... MITM services being created, that act as relays between clearnet users and the Tor network. So I'm not sure it would solve your concern about MITM services.

That being said, since the proposed approach would support Tor-only as well, the matter of whether or not to deprecate https is independent of what is being proposed here, and should probably be handled separately (even if both proposals may end up being implemented at the same time). You're welcome to make a proposal, I believe @j01tz has also been thinking along those lines so you might want to sync.

> By offering QR code and blob, you've now switched from a truly standardized format (JSON - currently used for all of our slates), to 2 different "standardized" formats

JSON is not being replaced by QR code or blob. We might end up serializing the slate in some other way than JSON, but it is the output of that serialization that will be encoded into a blob or QR code, or used in NFC, or whatever other ways of exchanging the 1st trip message that we come up with in the future.

It could also be that the blob ends up being what we encode into the QR code, and it might end up being what is passed around in all the other methods as well, replacing file for example.

Reply:
> How do services/users in regions where Tor is blocked transact?
> How do services/users that are prohibited culturally/legally from running Tor transact?

File

> How do mobile wallets transact?

Tor or file 

> How does the Grin network transact if/when Tor is down or unavailable?

In that extraordinarily rare scenario, file.

Reply:
> We cannot prevent users from running third party services, so that's definitely possible

But we can and should make predictions about what's likely to happen based on our experience over these past 14 months.

Reply:
`File`, `Tor`, or `Tor or File` as options implies that services and wallets need to give equal weight in their UIs to **two** transaction methods that are very different from each other, or serve a user different methods depending on which region they are connecting from, or from which device. One where you enter an address and transact, the other that's an  sequential: save / drag and drop file, save/export response, and upload response. 

Users need to learn both methods and form factors, increasing the cognitive load. 

The model proposed in this RFC has the identical experience regardless of whether Tor is supported or not by the service or end user. It can be carried out through the identical UI and flow, and is flexible to support multiple completion methods. 

Similar to how a typical bitcoin tx flow contains both an address in text and the same address as a QR code, I would expect services to provide the same slate as a QR code and a blob, and users would either copy/paste or scan with their phone, depending on which device they are on and on which device they have their wallet.

Services could provide both a clearnet Tor response URLs in the same slate payload, leaving it up the user to respond with whatever they can support on their end. 

It reduces friction significantly compared to requiring two distinctly different transaction flows.

Reply:
> leaving it up the user to respond with whatever they can support on their end.

And what if they can only support file? Then wallets and services still need to know how to support tor and file. I promise I'm not trying to be obtuse here, but I really can't quite grasp how this makes the situation any better at all.

Reply:
Under what circumstance would a user be able to interact with a service but not be able to support making an outbound https request?

They'd have to be offline, right? If they are offline, then I guess they are in the same physical location? If so, they might be able to initiate over NFC? That's still tbd as it seems like a minor use case for `service <> user` at this stage. 

For `user <> user`, they can still interact with file, file is not disappearing?

Reply:
Right, so for user <> user, wallets still need to support file. That's my point. Wallets still "need to give equal weight in their UIs to *two* transaction methods that are very different from each other"

Reply:
I see now what you mean. Yes, sure. *Wallets* would. They probably ought to support direct Tor transacting as well. *Services* on the other hand, would not. And that is the point of this RFC, to improve the `Service <> User` transaction flows, to the point of establishing a good  baseline. I don't see how offering `Tor`, or `Tor + File`, does this in a better way, and I don't understand how offering the hybrid method has an impact on the ability to transact using Tor. It's supported! 

I hint at `User <> User` flows and `File` in the unresolved questions section of the RFC. It's interesting to see how the armored slate idea will proceed, it might be that we could merge File and Blob, and then hybrid becomes an optional response mechanism, primarily intended to be used when transacting with services. Not sure.

Reply:
Tor only is easier to use and provides payment proofs. How would payment proofs work in your `tor + file` scenario?